feature/xops360-1385 Добавление FastAPI

stepan-im opened the pull request:
- добавление скрипта fast_api_service для обработки get и post HTTP запросов к RAG
- добавление Dockerfile и docker-compose для запуска fast-api-service
- реализация параллельного запуска сервиса и бота через .Dockerfile
- добавление весов модели 'intfloat/multilingual-e5-base', для исключения скачивания их из репозитория во время запуска образа

Commit:
Удаление файлов, которые не должны быть в GIT

Commit:
Обновление .gitignore

Commit:
Добавление FastAPI для работы с функционалом RAG

- добавление скрипта fast_api_service для обработки get и post HTTP запросов к RAG
- добавление Dockerfile и docker-compose для запуска fast-api-service

GRIMaxim commented:
Можешь удалить .idea и прочие папки в соответствии с .gitignore? трудно смотреть

GRIMaxim commented:
Поставь, пожалуйста, созвон, коментариев много

stepan-im commented:
> Можешь удалить .idea и прочие папки в соответствии с .gitignore? трудно смотреть

Так я их и удалил.

Commit:
Рефакторинг fast-api-service по предложениям Максима

- Корректировка запуска fast_api_service через uvicorn и открытие порта 8000 в Dockerfile
- Добавление пробрасывания порта в compose.yaml
- Добавление ограничения поля body в Message
- Корректировка обработки команд в fast_api_service.py

GRIMaxim commented on fast_api_service.py:
```diff
@@ -0,0 +1,43 @@
+from programs import AVAILABLE_PROGRAMS
+from fastapi.encoders import jsonable_encoder
+
+app = FastAPI(prefix="/api", description=f"Это сервис, который отвечает на вопросы по документам по господдержке. "
```
Обычно пишут в духе prefix=/api/v1/, но не критично

stepan-im replied:
Изменил

GRIMaxim commented on fast_api_service.py:
```diff
@@ -0,0 +1,43 @@
+    if program_number == '-1':
+        return JSONResponse(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            content=jsonable_encoder(Message(body="Пока реализация такова, "
```
Слишком сложно и в документации потом отражать будет трудно
https://fastapi.tiangolo.com/ru/tutorial/handling-errors/
В соответствии с этим замени return JSONResponse на raise HTTPException, если мы говорим об ошибках
Добавь в 20 строку параметр status_code=HTTP_200
В 40 строке просто введи return {'body': f"{message} {response}"}

stepan-im replied:
Поправил.

GRIMaxim commented on fast_api_service.py:
```diff
@@ -0,0 +1,43 @@
+
+@app.get("/healthcheck", status_code=status.HTTP_200_OK, response_model=Message, summary='Проверка работоспособности сервиса')
+async def healthcheck():
+    return Message(body="Сервис работает")
```
Здесь просто return {'message': 'OKResponce'}, в 15 строке responce_model можно убрать

stepan-im replied:
Скорректировал содержание ответа, структуру предлагаю оставить прежней

GRIMaxim commented on FastAPI.Dockerfile:
```diff
@@ -0,0 +1,13 @@
+
+COPY . .
+
+EXPOSE 8000
```
Это лишнее, порт и так прокидывается в docker-compose

Commit:
Help for Stepan-Docker

GRIMaxim commented on fast-api-docker-compose.yaml:
```diff
@@ -0,0 +1,13 @@
+      context: .
+      dockerfile: FastAPI.Dockerfile
+    volumes:
+      - ~/rag_serice/log/:/var/log/
```
/var/log/rag_service/log/

GRIMaxim replied:
изменений не вижу

Commit:
Правки по комментариям Максима + обновление README.md

GRIMaxim commented on fast_api_service.py:
```diff
@@ -0,0 +1,39 @@
+@app.get("/healthcheck", status_code=status.HTTP_200_OK, response_model=Message,
+         summary='Проверка работоспособности сервиса')
+async def healthcheck():
+    return "OKResponce"
```
{"detail": "OKResponce"}

GRIMaxim commented on fast_api_service.py:
```diff
@@ -0,0 +1,39 @@
+    rag = RAG(query=message.body)
+    program_number = rag.get_program_number(query=message.body)
+    if program_number == '-1':
+        return HTTPException(
```
raise вместо return

stepan-im replied:
Поправил

GRIMaxim commented on fast_api_service.py:
```diff
@@ -0,0 +1,39 @@
+                   "Введите вопрос с номером постановления"
+        )
+    elif program_number not in AVAILABLE_PROGRAMS:
+        return HTTPException(
```
raise вместо return

stepan-im replied:
Поправил

GRIMaxim commented on fast_api_service.py:
```diff
@@ -0,0 +1,39 @@
+    else:
+        message = f"Постановления, в которых буду искать ответ на вопрос: {program_number}."
+        response = await rag.process()
+        return f"{message} {response}"
```
return {"body": f"{message} {response}"}

stepan-im replied:
Обговорили голосом

Commit:
Правки по комментариям Максима

Commit:
Корректировка параметра FastAPI и проброс логов в Dockerfile

Commit:
Корректировка корневого пути для ссылок запросов

Commit:
Формализация ответов post и get в fast_api_sevice

Commit:
Merge remote-tracking branch 'origin/e5_pipeline' into feature/XOPS360-1385-add-fast-api

# Conflicts:
#	RAG/__pycache__/llm_ident.cpython-311.pyc
#	RAG/__pycache__/prompts.cpython-311.pyc
#	RAG/__pycache__/rag.cpython-311.pyc

Commit:
Отладка fast_api_service

Commit:
Отладка parse_question(...)

Commit:
Добавление compose и Dockerfile файлов для запуска Бота и сервиса в одном контейнере

polIINNa commented on fast_api_service.py:
```diff
@@ -0,0 +1,39 @@
+    program_number = rag.get_program_number(query=message.body)
+    if program_number == '-1':
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
```
В этом случае не должна быть ошибка 404. Если получилось -1, API все равно должно быть доступно. Бот будет получать сообщение "Пока реализации такоова..."  и выводить его пользователю

GRIMaxim replied:
Договорились на 200 статус

stepan-im replied:
Поправлю

stepan-im replied:
Хотя, как по мне, пользователь должен определять, что он не прав не по анализу сообщения, а через стандартные коды ошибок.

GRIMaxim replied:
Ну тут смысл в том, что 400 ошибка относится больше к запросам, чем к результатам работы RAG
Такие сообщения будут лететь клиенту сразу, а вот 4хх оставим на более серьезные вещи

stepan-im replied:
Поправил

polIINNa commented on fast_api_service.py:
```diff
@@ -0,0 +1,39 @@
+                   "Введите вопрос с номером постановления"
+        )
+    elif program_number not in AVAILABLE_PROGRAMS:
+        raise HTTPException(
```
Здесь аналогичный момент: если program_number не в AVALIABLE, клиент получает соответствующее сообщение и выводит его пользователю. Тут не должна падать ошибка

GRIMaxim replied:
200 статус

stepan-im replied:
Поправлю

GRIMaxim replied:
Здесь забыл, вместо Exception Message(body=f"{caption} {response}")

GRIMaxim replied:
и поменять местами {responce} и {caption} местами
а caption имеет формат {Документ, в котором был найден ответ: 1570}

stepan-im replied:
> Здесь забыл, вместо Exception Message(body=f"{caption} {response}")

Это поправил

stepan-im replied:
> и поменять местами {responce} и {caption} местами а caption имеет формат {Документ, в котором был найден ответ: 1570}

Это не понял.

Commit:
Merge branch 'master' into feature/XOPS360-1385-add-fast-api

# Conflicts:
#	.gitignore

Commit:
Доработка параллельного запуска сервиса и бота, корректировка по комментариям

- Реализация параллельного запуска сервиса и бота через .Dockerfile
- Корректировка кодов ответов в parse_question(...)

Commit:
Merge branch 'master' into feature/XOPS360-1385-add-fast-api

GRIMaxim commented on fast_api_service.py:
```diff
@@ -0,0 +1,39 @@
+    rag = RAG()
+    program_number = rag.get_program_number(query=message.body)
+    if program_number == '-1':
+        raise HTTPException(
```
Здесь забыл, вместо Exception Message(body=f"{caption} {response}")

GRIMaxim replied:
и поменять местами {responce} и {caption} местами
а caption имеет формат {Документ, в котором был найден ответ: 1570}

stepan-im replied:
> Здесь забыл, вместо Exception Message(body=f"{caption} {response}")

Это поправил

stepan-im replied:
> и поменять местами {responce} и {caption} местами а caption имеет формат {Документ, в котором был найден ответ: 1570}

Это не понял

Commit:
Правка fast_api_service по комментариям Максима

GRIMaxim commented on fast_api_service.py:
```diff
@@ -0,0 +1,33 @@
+    elif program_number not in AVAILABLE_PROGRAMS:
+        return Message(body=f"Таких постановлений нет в базе. Вот список доступных: {AVAILABLE_PROGRAMS}")
+    else:
+        caption = f"Постановления, в которых буду искать ответ на вопрос: {program_number}."
```
Вот тут поменять caption=f"Документ, в котором был найден ответ: {program_number}"

GRIMaxim commented on fast_api_service.py:
```diff
@@ -0,0 +1,33 @@
+    else:
+        caption = f"Постановления, в которых буду искать ответ на вопрос: {program_number}."
+        response = await rag.process(query=message.body)
+        return Message(body=f"{caption} {response}")
```
Вот тут Message(body=f"{response}\n{caption}")

GRIMaxim replied:
и все, дальше approve, вливаем и полина будет правки по рагу делать

stepan-im replied:
Почему наоборот?
Сначала пишем, где ищем, а потом ответ модели. Также сделано в боте

GRIMaxim replied:
Это от Алины правка

stepan-im replied:
Поправил

Commit:
Корректировка формата вывода ответа от модели в fast_api_service
